internal/engine/compiler/engine.go

```diff
@@ -1151,8 +1151,16 @@ func (ce *callEngine) builtinFunctionMemoryGrow(mem *wasm.MemoryInstance) {
 
 	// Update the moduleContext fields as they become stale after the update ^^.
 	bufSliceHeader := (*reflect.SliceHeader)(unsafe.Pointer(&mem.Buffer))
-	ce.moduleContext.memorySliceLen = uint64(bufSliceHeader.Len)
-	ce.moduleContext.memoryElement0Address = bufSliceHeader.Data
+	if mem.Shared {
+		// Use atomic to ensure visibility for good measure. Though in practice, we know
```

Comment:
This I just found when comparing the two repos. As per this comment, it's likely not a big deal but kept it just to have a "best practice" of always using atomic writes in shared mode, I can revert if not needed though

Reply:
I'm getting a bit uneasy about all these atomic stores that don't seem to have corresponding atomic loads.

I know the loads are in assembly, and the amd64 versions are probably safe (atomic stores have an mfence, or xchg instruction, loads don't, apparently) but I'm not sure about arm64.

Did you see how Go implements atomic, on godbolt or similar?

Reply:
I had assumed Go's implementation but indeed hadn't verified it. Just checked, luckily they're pretty straight forward

https://github.com/golang/go/blob/adead1a93f472affa97c494ef19f2f492ee6f34a/src/runtime/internal/atomic/atomic_arm64.s#L120

STLR, same as what we use

https://github.com/tetratelabs/wazero/blob/7d5b6739da3ae7456db41399a2701e22d66b57c4/internal/engine/compiler/impl_threads_arm64.go#L127

For reading with its sister instruction, LDAR

https://github.com/tetratelabs/wazero/blob/7d5b6739da3ae7456db41399a2701e22d66b57c4/internal/engine/compiler/impl_arm64.go#L4052

Does it seem ok?

Reply:
For good measure, amd64, store is xchg as we do

https://github.com/golang/go/blob/adead1a93f472affa97c494ef19f2f492ee6f34a/src/runtime/internal/atomic/atomic_amd64.s#L174

and load is just normal load same as we do too

https://github.com/golang/go/blob/adead1a93f472affa97c494ef19f2f492ee6f34a/src/runtime/internal/atomic/atomic_amd64.go#L30

Reply:
Right, ignore me, I guess I missed the A in `LARD` [here](https://github.com/tetratelabs/wazero/blob/6166dae7410790a087487063384c501639ec2139/internal/engine/compiler/impl_arm64.go#L4052).